fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,560 @@
+        }
+        ++tokenIt; // consume to be the start of the next Arg (could be single node value or ID followed by BECOMES)
+    }
+    return argList;
+}
+
+/* Parse and build the AST
+ * 
+ * @param tokenIt The tokens contains the AST info
+ * @return The root of the AST
+ */
+ASTNode* buildAST(TokenIter &tokenIt) {
+    ASTNode* siblingsList = NULL;
+    while (!tokenIt.isEnd() && tokenIt->getType() == Token::LPAREN) {
+        ASTNodeArg* argList = NULL;
+        ASTNode* currentchild = NULL;
+        ASTNode* childrenList = NULL;
+        ++tokenIt; // consume to be ID
+        if (tokenIt.isEnd()) {
+            throw ParserFailure("Token stream ended unexpectedly after LPAREN");
+        } else if (tokenIt->getType() != Token::ID) {
+            throw ParserFailure("The next token of LPAREN should be IDENTIFIER, while current next value is " +  tokenIt.peek().getValue());
+        }
+        char* node_name = new char[tokenIt->getValue().size() + 1];
+        strcpy(node_name, tokenIt->getValue().c_str());
+        ++tokenIt; // consume to be the start of argList/nodeList or RPAREN
+        if (checkValue(tokenIt)) {
+            try {
+                argList = parseArgList(tokenIt);
+            } catch (LexerFailure &f) {
+                std::cerr << f.what() << std::endl;
+            }
+        }
+        currentchild = buildAST(tokenIt);
+        if (childrenList == NULL) {
+            childrenList = currentchild;
+        } else {
+            appendSiblingNode(childrenList, currentchild);
+        }
+        if (!tokenIt.isEnd() && tokenIt->getType() != Token::RPAREN) {
+            throw ParserFailure("Missing RPAREN, while current value is " + tokenIt->getValue());
+        }
+        ++tokenIt; // consume to be the end of file or LPAREN(the sibling of the current node)
+
+        if (siblingsList == NULL) {
+            siblingsList = createNode(node_name, argList, childrenList, NULL);
+        } else {
+            appendSiblingNode(siblingsList, createNode(node_name, argList, childrenList, NULL));
+        }
+    }
+    return siblingsList;
+}
+
+/* Parse the file to build AST
+ * 
+ * @param in The tril input file
+ * @return The root of the AST
+ */
+ASTNode* parseFile(FILE *in) {
+    int c;
+    std::string str;
+    while ((c = std::fgetc(in)) != EOF) {
+        str += c;
+    }
+    std::vector <Token> scanToken = scan(str);
+    TokenIter token = TokenIter(scanToken.begin(), scanToken.end());
+    ASTNode* head = NULL;
+    try {
+        head = buildAST(token);
+    } catch (ParserFailure &f) {
+        std::cerr << f.what() << std::endl;
+    }
+    return head;
+}
+
+/* Parse the string to build AST
+ * 
+ * @param in The whole tril input
+ * @return The root of the AST
+ */
+ASTNode* parseString(const char* in) {
+    std::vector <Token> scanToken;
+    std::string in_str(in);
+    std::vector <Token> tokenLine = scan(in_str);
+    scanToken.insert(scanToken.end(), tokenLine.begin(), tokenLine.end());
```

Comment:
What's the point of having both `scanToken` and `tokenLine`? I think you should only need one of these.